core/trino-main/src/main/java/io/trino/execution/scheduler/SqlQueryScheduler.java

```diff
@@ -297,6 +297,9 @@ else if (state == QueryState.FAILED) {
         if (queryStateMachine.isDone()) {
             return Optional.empty();
         }
+        if (attempt > 0 && retryPolicy == RetryPolicy.QUERY) {
```

Comment:
Where in query flow `createDistributedStagesScheduler` is being executed? Are we certain there won't be any communication/DFs received for previous attempt after `registerQueryRetry` is called?

Reply:
This should happen after stages scheduler of previous attempt has gone into failed state. I am assuming that RemoteTask of previous attempt are cleaned up before a retry is scheduled, so there shouldn't be any fetching of DF from last attempt. @arhimondr please confirm if this is correct assumption.

Reply:
That's a very good point @sopel39 

The scheduling and task updates happen asynchronously, thus the process is inherently racy. While the retry happens with a delay getting a delayed update is rather unlikely, but possible. Ideally `DynamicFilterService` should reject updates from tasks from previous attempts.

Reply:
I've updated the code now to ignore updates from previous attempts.